Add token, user, package model tests

olingern opened the pull request:
I thought it would be nice to get to know the project and write some of the registry's model tests. 

Related issue #83

chrisdickinson commented:
Thanks for the PR! Ideally I'd like to use the `providePostgres` helper for all tests that touch the database: the goal of the helper is not just to provide the connection, but also to run all modifications to the database during a testcase in a transaction which is rolled back at test completion. This pattern (borrowed from Django) isolates test state between test cases. It'd definitely be worthwhile to look into allowing the connection to be reused between test cases, though!

olingern commented:
@chrisdickinson got these passing :tada: 

- Had to add the env vars to the circle ci config
- Add mocha to the eslintrc

olingern commented:
@chrisdickinson to the comment above:

`the goal of the helper is not just to provide the connection, but also to run all modifications to the database during a testcase in a transaction which is rolled back at test completion`

So, I read this and the `providePostgres` function as:
- Start with a clean slate every individual test
- Wrap everything in a transaction

By using `beforeEach` we're saying, _"Open a transaction every test"_ and using `afterEach` rollback the previous transaction.

I do like the `providePostgres` function, but (_currently_) it doesn't allow errors to propagate up and it's difficult to capture results inside the passed function scope and test them outside of it. My specific use case is capturing the errors that Joi / ormnomnom throws. I'm not tied to this. If there's a better way to do with within the confines of `provideRegistry`'s implementation, I can do that :slightly_smiling_face:

zacanger commented:
@olingern can you pull from upstream master? I think some of the things you're fixing are fixed already.

olingern commented:
@zacanger Sure, done. I only saw one conflict which was with d54ab504c146ef327cf5d209e3a2aded4cdfb72a, so these should be okay

olingern commented:
@chrisdickinson refactored per your comment above. Also moved this under services. Let me know if you'd like me to drop b5b681dae3a24258b170e5cabba15ef1f0c018ec per the conversation in #133

olingern commented:
Will squash / fixup these commits after review :slightly_smiling_face:

olingern commented on services/registry/models/token.js:
```diff
@@ -53,6 +52,10 @@ module.exports = class Token {
   }
 
   static async lookupUser(value) {
+    if (!value || value.indexOf('ent_v1_') !== 0) {
```
Sanity check added here.

chrisdickinson replied:
🎉 This is a great idea! (A tiny nit, but could we switch from `indexOf` to `startsWith`?)

olingern replied:
Sure, updated

olingern commented on services/registry/models/token.js:
```diff
@@ -1,6 +1,5 @@
 'use strict';
 
-const querystring = require('querystring');
```
I saw this wasn't used

olingern commented on services/registry/test/models/user.js:
```diff
@@ -0,0 +1,173 @@
+      'should create a remoteAuth when remoteAuth param is provided',
+      providePostgres(async () => {
+        // Kind of dirty, but we currently only need it for this test
+        let prevOAuthEnvVar = process.env.OAUTH_PASSWORD;
```
Like my comment says, it's not the best. Altering `process.env` feels weird ... but I think we would need some sort of global setup script that would load something like a `.env-test` file to avoid this

jacques replied:
Better to use the .env-test file for this.

chrisdickinson replied:
I don't mind messing with the environment here. I might suggest doing this bit in a separate "before / after" pair to make sure it's always run.

olingern replied:
> I might suggest doing this bit in a separate "before / after" pair to make sure it's always run.
Done!

olingern commented on services/registry/test/models/package.js:
```diff
@@ -0,0 +1,52 @@
+
+const { createUser } = require('../utils/users');
+
+async function getUserNamespace(name, email) {
```
This might should be moved depending on how "helpful" it is

olingern commented on .eslintrc:
```diff
@@ -1,12 +1,14 @@
     "node": true,
-    "es6": true
+    "es6": true,
+    "mocha": true
```
Added `mocha` option due to getting `described` and `it` are undefined on the CI.

olingern commented:
Rebased master that has #154 in it, and now passing :tada: 

Cleaned this up and keeping it to only token, user, and package tests

chrisdickinson commented on services/registry/test/models/package.js:
```diff
@@ -0,0 +1,40 @@
+  before(async () => {});
+
+  it(
+    'Given minimum, valid paramters it creates a new package ',
```
typo: `parameters`

olingern replied:
Fixed!

chrisdickinson commented on services/registry/test/models/user.js:
```diff
@@ -0,0 +1,172 @@
+        demand(remoteAuthObjects[0].active).to.be.truthy();
+
+        // reset
+        process.OAUTH_PASSWORD = prevOAuthEnvVar;
```
Typo: `process.env.OAUTH_PASSWORD` (missing `.env`)

olingern replied:
:eyes:

olingern commented:
@chrisdickinson thanks for the review! I've been having some problems with testing the few nits, so I opened #190

olingern commented:
@chrisdickinson I went through each nit and updated and also:
- Updated the `npm test` command to use a glob
- Renamed the test files to `test.spec.js` ( _this allows us to easily exclude utils instead of a complex regex_ )
- Moved the integration tests under `integration`

If those changes are :+1: I'll do some commit squashing

chrisdickinson commented on services/registry/test/models/user.spec.js:
```diff
@@ -0,0 +1,178 @@
+
+        // Fetch the user's namespace that was created during signup.
+        const userNamespace = await Namespace.objects.filter({
+          'host.id': 1, // TODO: don't hardcode?
```
I'd be tempted to add a helper function to `Host` to grab the current hostname so we can get condense all of the `process.env.EXTERNAL_HOST.replace(/https?:\/\//, '')` bits into a single `Host.current()` call. That would let us write `'host.name': Host.current()` here, for example!

(This can definitely be left for a separate PR.)

olingern replied:
Oh, that would be nice! When I first looked at this, I had to do some logging to figure out what 1 actually meant. So a helper would be ... helpful :slightly_smiling_face:

olingern commented:
Alright, rebase done. Moved the previous registry tests into storage and updated accordingly. I should note that `pretest` fails on my machine. Can't connect to the postgres instance. Removing `pretest` and then running `npm t` inside of the storage service works, though

chrisdickinson commented:
Thanks for working on this! I'm going to plan on merging this tomorrow AM.

chrisdickinson commented:
Merged in 7f935831f1da132a92e964b36ef3434b92223575, thanks!

olingern commented:
awesome :rocket: !
